### PR TITLE
Don't require fake ammo type for UPS-only weapons

### LINKED ISSUE #15128: Cannot modify UPS Weapons without unloading them first. Cannot unload UPS weapons. 0.C-9984-ga87baf6
Steps to reproduce: 
Spawn in an A7 laser rifle and your choice of UPS (item or cbm), and a shoulder strap.
Power the UPS
Fire the rifle once
'A'ctivate the shoulder strap and select the A7 laser rifle as the target.
You will get the message: "Unload your A7 laser rifle before trying to modify it."
https://i.imgur.com/pfgmQ7y.png
Alternatively, spawn in the laser rifle and the shoulder strap, attach the shoulder strap in the normal fashion and THEN fire the weapon. Try 'A'ctivating the laser rifle to remove the mod and you will recieve the same message.

Before firing, there is no ammuntion present in the line below the ammo specification: https://i.imgur.com/ZPrtkiF.png

After firing, placeholder ammunition is present: https://i.imgur.com/J4zm011.png

The weapon CANNOT be unloaded: https://i.imgur.com/ENJUP18.png

Some kind of method for bypassing the unloaded-only modification restriction for UPS powered weapons is needed.

### PR DESCRIPTION
**Pull request by mugling**
This PR deprecates `generic_no_ammo` whilst preserving support for weapons that consume only UPS charges.
#### Behavioral changes
- `item::info` now displays UPS usage for guns specifying `ups_charges`
- This includes any acquiring the value via attached gunmods
#### Implementation details
- Deprecates `has_curammo` and `get_curammo` accessors via migration to `ammo_data()`
- Updates all consumers of `ammo_data()` to be tolerant of it returning `nullptr`
- Provides `ammo_effects` and `ammo_casing` acessors (also improves encapsulation)
- `vehicle::*_fire_turret` now accepts the gun as an argument
- Forbids specifying both `ammo` and `NO_AMMO` in json data
#### Outstanding issues
- Fixes #15128
- Charger guns already broken and remain so (fix in separate PR)

### CODE REVIEW

File: src/item.cpp [hunk 1]
```diff
@@ -4395,22 +4423,12 @@ item_location item::pick_reload_ammo( player &u ) const
...
-    // Casings need a count of one to stack properly.
-    item casing( curammo->ammo->casing, calendar::turn );
-    casing.charges = 1;
-
-    // Drop all the casings on the ground under the player.
-    for( auto i = target.get_var( "CASINGS", 0 ); i > 0 ; --i ) {
-        g->m.add_item_or_charges( p.posx(), p.posy(), casing );
-    }
-
+    g->m.add_item_or_charges( p.posx(), p.posy(), item( target.ammo_casing(), calendar::turn, qty ) );
```
Comment (Coolthulhu) [inline src/item.cpp]: This should have the 3D position argument, otherwise npcs will drop casings on player's z-level.

File: src/ranged.cpp [hunk 2]
```diff
@@ -1222,28 +1222,27 @@ static projectile make_gun_projectile( const item &gun ) {
...
 
-    const auto curammo = gun.ammo_data();
-
-    // Consider both effects from the gun and ammo
     auto &fx = proj.proj_effects;
-    fx.insert( gun.type->gun->ammo_effects.begin(), gun.type->gun->ammo_effects.end() );
-    fx.insert( curammo->ammo->ammo_effects.begin(), curammo->ammo->ammo_effects.end() );
 
-    if( curammo->phase == LIQUID || fx.count( "SHOT" ) || fx.count("BOUNCE" ) ) {
+    if( ( gun.ammo_data() && gun.ammo_data()->phase == LIQUID ) ||
```
Comment (Coolthulhu) [inline src/ranged.cpp]: It may be a good idea to have a default `ammo_data` now, to avoid checking for `nullptr`.

File: src/item.cpp (hunk 1 above)
Comment (mugling) [inline src/item.cpp]: Fixed in b982c89

File: src/ranged.cpp (hunk 2 above)
Comment (mugling) [inline src/ranged.cpp]: That could work. I'll look into it in a separate PR

**Pull request merged**